Murmur3 iobuf

bashtanov opened the pull request:
Implement murmur3 hash for iobuf.

If buffers are oddly sized or allocated we access uint32_t numbers using a pointer which is not a multiple of 4 bytes. It is okay on x86 and aarch64 but not on some other architectures.

It'll be needed for bucketing strings in iceberg transformations

microbench results:
```
test                          iterations      median         mad         min         max      allocs       tasks        inst      cycles
murmur_hash_x86_32.aligned       4824000   203.608ns     0.130ns   203.018ns   203.961ns       0.000       0.000      2800.3      1115.3
murmur_hash_x86_32.unaligned     4827000   207.595ns     0.380ns   206.494ns   208.470ns       0.000       0.000      2798.8      1137.6
murmur_hash_x86_32.iobuf         4918000   202.882ns     0.079ns   202.637ns   203.302ns       0.000       0.000      3333.3      1111.5
```

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
hashing/tests/murmur: murmurhash3_x86_32 consistency for iobuf vs. char*

bashtanov committed:
bazel: sanitize for unaligned access

vtools cmake does it in debug mode

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a4e16d99-cbd4-4fc9-8ca9-973f3ce9748a-->
<!--TEST RESULTS-->
<!--06cd8bd2-2396-4a9b-bb88-58148a4c2991-->
<!--TEST RESULTS-->
<!--76306b13-3c34-4a98-9c32-9f8cbb193101-->
<!--TEST RESULTS-->
<!--ef477fa0-0633-4593-a2bf-1c85e4fb1606-->
<!--TEST RESULTS-->
<!--3cb85d89-4ae7-4feb-97ea-b6454125e5ec-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63086">build#63086</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63086#0195919a-4198-471c-afca-3241e7132c5a | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63086#0195919a-419a-48ba-847f-e82505f34cd4 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63086#0195919a-4198-471c-afca-3241e7132c5a | FLAKY | 1/3 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63086#0195919a-4199-4c1f-91b2-85c5ab6329ad | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move_x_core.replication_factor=3.unclean_abort=False.recovery=restart_recovery.compacted=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63086#0195919a-4199-4c1f-91b2-85c5ab6329ad | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63149">build#63149</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/63149#01959520-094b-46a2-8783-d7fe38b4c49e | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63149#01959563-2a3b-473c-ac0e-9f37e4a98aa9 | FLAKY | 1/2 |
| rptest.tests.write_caching_fi_test.WriteCachingFailureInjectionTest.test_unavoidable_data_loss | ducktape | https://buildkite.com/redpanda/redpanda/builds/63149#01959563-2a3a-472f-b6a4-a0c589856e83 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63226">build#63226</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63226#0195a482-1f2c-4980-8975-be6ee3a0b1e4 | FLAKY | 1/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_spec_evolution.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63226#0195a482-1f2c-4980-8975-be6ee3a0b1e4 | FLAKY | 1/2 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/63226#0195a482-1f2c-4980-8975-be6ee3a0b1e4 | FLAKY | 1/2 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63226#0195a467-4052-475e-8382-3be7efda02f8 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63262">build#63262</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/63262#0195a65a-6c4d-4f68-a05b-82ad198bef60 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63262#0195a699-c225-4a36-8ab3-713b48666b4f | FLAKY | 1/3 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63262#0195a699-c226-4def-b2bc-f2d5f8d0dfb3 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63281">build#63281</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63281#0195a8a9-2f35-4ca3-9121-0afa203d72e6 | FLAKY         | 1/2      |
</details>
<!--/body-->

bashtanov committed:
hashing/tests/murmur3: benchmark

make sure it's not significantly slower when unaligned

WillemKauf commented on .bazelrc:
```diff
@@ -54,8 +54,8 @@ build --@seastar//:debug=True
 # =================================
-build:sanitizer --copt -fsanitize=address,undefined,vptr,function
-build:sanitizer --linkopt -fsanitize=address,undefined,vptr,function
+build:sanitizer --copt -fsanitize=address,undefined,vptr,function,alignment
```
nice

WillemKauf commented on src/v/hashing/murmur.cc:
```diff
@@ -109,6 +116,97 @@ uint32_t murmurhash3_x86_32(const void* key, std::size_t len, uint32_t seed) {
+    // body
+
+    uint32_t torn_block;
+    auto torn_block_begin = reinterpret_cast<char*>(&torn_block);
```
There is a lot more going on here than in the `murmurhash3_x86_32(const void* key, std::size_t len, uint32_t seed)` implementation. Can you leave some comments explaining the additional work/this additional torn block variable?

WillemKauf commented on src/v/hashing/tests/hash_bench.cc:
```diff
@@ -254,3 +258,41 @@ PERF_TEST(ntp_hash, absl_hash_of) {
+    return murmurhash3_x86_32_body<0>();
+}
+
+PERF_TEST(murmur_hash, murmurhash3_x86_32_unaligned) {
```
>make sure it's not significantly slower when unaligned

Any results from the perf test you can share? Did you benchmark the `iobuf` implementation as well?

bashtanov replied:
```
test                          iterations      median         mad         min         max      allocs       tasks        inst      cycles
murmur_hash_x86_32.aligned       4824000   203.608ns     0.130ns   203.018ns   203.961ns       0.000       0.000      2800.3      1115.3
murmur_hash_x86_32.unaligned     4827000   207.595ns     0.380ns   206.494ns   208.470ns       0.000       0.000      2798.8      1137.6
murmur_hash_x86_32.iobuf         4918000   202.882ns     0.079ns   202.637ns   203.302ns       0.000       0.000      3333.3      1111.5
```

dotnwat commented on src/v/hashing/murmur.cc:
```diff
@@ -109,6 +116,97 @@ uint32_t murmurhash3_x86_32(const void* key, std::size_t len, uint32_t seed) {
+    auto torn_block_data_end = torn_block_begin;
+
+    for (const auto& fragment : data) {
+        auto frag_begin = fragment.get();
```
it seems like the body of this for-loop should be some kind of generic non-iobuf-specific interface that the iobuf adapter reuses.

dotnwat replied:
it seems like the body of this for-loop should be some kind of generic non-iobuf-specific interface that the iobuf adapter reuses.

also please explain more in the commit message how you built this. is it an adaptation of existing implementation? what's unique about it and why aren't we using existing murmur3 interfaces? what does it mean for the byte code to be identical? etc...

dotnwat commented on src/v/hashing/tests/murmur_tests.cc:
```diff
@@ -0,0 +1,97 @@
+}
+} // namespace
+
+BOOST_AUTO_TEST_CASE(iobuf_equivalent_to_continuous) {
```
new tests should be gtest when possible

bashtanov committed:
coding style: prefer gtest

bashtanov committed:
hashing/murmur: move internal functions to an anonymous namespace

bashtanov committed:
hashing/murmur: use std::memcpy to legalize unaligned access

also use reinterpret_cast instead of c-style casts

bashtanov committed:
hashing/murmur: refactor x86_32 murmur3

Separate block, tail and final stages to reuse them in later when
implementing for non-continuous input (iobuf)

bashtanov committed:
hashing/murmur: implement x32_86 for iobuf

Handle wrapped blocks correctly, still benefit from continuous blocks
and series of them within the fragment

bashtanov committed:
hashing/tests/bench/murmur3: benchmark iobuf hashing

bashtanov commented:
@dotnwat @WillemKauf I restructured hash calculation. On godbolt (clang18 -O3 x86-64) the old and the new implementations for continuous span result into nearly-identical ASM code (only uses different registers in a few places).

bashtanov committed:
bytes/random: convert a string view to a randomly split iobuf

To test iobuf algorithms

mmaslankaprv commented on src/v/bytes/random.h:
```diff
@@ -22,4 +24,6 @@ bytes get_crypto_bytes(size_t n = 128 * 1024, bool use_private_rng = false);
 // Makes an random alphanumeric string, encoded in an iobuf.
 iobuf make_iobuf(size_t n = 128);
 
+iobuf split_as_iobuf(std::string_view str, int n_fragments);
```
I think this should go into a different header

bashtanov replied:
any preference where about?
or maybe would it look better here if it accepted `bytes` instead of `std::string_view`?

mmaslankaprv replied:
it is not related with `random` at all, maybe utils ?

bashtanov replied:
It is, as it chooses fragment sizes randomly.

mmaslankaprv replied:
i missed that, let's leave it as is then

Lazin commented on src/v/hashing/tests/hash_bench.cc:
```diff
@@ -254,3 +259,49 @@ PERF_TEST(ntp_hash, absl_hash_of) {
+}
+
+template<size_t... Offset>
+size_t murmurhash3_x86_32_continuous() {
```
can you add benchmark output to the commit message or to the PR description?

bashtanov replied:
added to PR description

Lazin commented on src/v/hashing/murmur.cc:
```diff
@@ -129,7 +133,7 @@ void murmurhash3_x86_128(
+    auto blocks = reinterpret_cast<const uint32_t*>(data + nblocks * 16);
 
     for (int i = -nblocks; i; i++) {
         uint32_t k1 = internal::getblock32(blocks, i * 4 + 0);
```
I guess the main reason why unaligned input doesn't affect performance is that it always does unaligned reads anyway.

bashtanov replied:
Do you mean it does not generate specific optimized code for the aligned case? It is the case on godbolt with minimal compiler flags, and I would assume it's the same in RP.

Lazin replied:
Yes, it can't be vectorized. It reads data using `uint32_t*` but on every step it does 4 reads moving pointer forward by one byte every time. So no matter the alignment we're getting one aligned read and three misaligned.

Lazin commented on src/v/hashing/murmur.cc:
```diff
@@ -12,7 +12,7 @@
-namespace internal {
+namespace {
 inline uint32_t rotl32(uint32_t x, int8_t r) {
     return (x << r) | (x >> (32 - r));
```
the code is valid in C but it's an UB because x could be shifted by 32 places to the right (if r is 0) or to the left if r > 32
should be
```
assert(r > 0);  // normal assert that will be removed in release
r &= 31;
return (x << r) | (x >> (32 - r));
```

Lazin replied:
I guess it's not important because `r` is always a constant with proper value

WillemKauf commented on src/v/hashing/murmur.cc:
```diff
@@ -102,13 +87,35 @@ uint32_t murmurhash3_x86_32(const void* key, std::size_t len, uint32_t seed) {
+        x86_32::consume_block(h1, k1);
+    }
+
+    const uint8_t* tail = (const uint8_t*)(data + nblocks * 4);
```
nit: also meant to be a `reinterpret_cast()`?

dotnwat commented on src/v/bytes/random.cc:
```diff
@@ -37,4 +38,25 @@ iobuf make_iobuf(size_t n) {
     return io;
 }
 
+iobuf split_as_iobuf(std::string_view str, int n_fragments) {
+    auto remaining_len = str.length();
+    std::vector<size_t> fragment_lengths;
+    fragment_lengths.reserve(n_fragments);
+    for (int i = 0; i < n_fragments - 1; ++i) {
+        size_t fragment_len = get_int(remaining_len);
+        fragment_lengths.push_back(fragment_len);
+        remaining_len -= fragment_len;
+    }
+    fragment_lengths.push_back(remaining_len);
+    std::ranges::shuffle(fragment_lengths, internal::gen);
+
+    iobuf res;
+    auto cur = str.begin();
+    for (auto fragment_len : fragment_lengths) {
+        res.append(ss::temporary_buffer<char>(cur, fragment_len));
+        cur += fragment_len;
+    }
+    return res;
```
this should probably go into a random-specific test utility/library. i can't imagine this being useful in production right?

bashtanov replied:
By the function right above, which is also only used in tests, I judged that this library is a good place for this kind of function. Do you think we should move both to, say, `//src/v/random:generators`? It's used in production code too BTW, we don't seem to have a test-only library for randomization.

dotnwat replied:
> It's used in production code too BTW

where is it used in production? i only see these non-production usages

```
nwatkins@acid:~/src/redpanda$ git grep split_as_iobuf
src/v/bytes/random.cc:iobuf split_as_iobuf(std::string_view str, int n_fragments) {
src/v/bytes/random.h:iobuf split_as_iobuf(std::string_view str, int n_fragments);
src/v/hashing/tests/hash_bench.cc:    iobuf split_buf = random_generators::split_as_iobuf(
src/v/iceberg/tests/transform_utils_test.cc:        auto split = random_generators::split_as_iobuf(
src/v/iceberg/tests/transform_utils_test.cc:        auto split = random_generators::split_as_iobuf(
```

why would we need to randomly split a iobuf in a non-test situation?

bashtanov replied:
I meant `//src/v/random:generators` is used in production.

dotnwat replied:
right, so you can move the random splitting utility into a test-only library

dotnwat replied:
as an example, we have `test_utils/randoms.h` -- not that this should go there, i'm just highlighting that there is precedence.

bashtanov replied:
> not that this should go there

Why not? Especially if we combine it with data generation and call it `random_iobuf`.

dotnwat replied:
i'm not saying it shouldn't go there, i'm just saying it's an example of a place it could go.
